tabular/src/autogluon/tabular/models/widedeep_nn/widedeep_nn_model.py

```diff
@@ -0,0 +1,289 @@
+                X_train=X_train,
+                X_val=X_valid,
+                n_epochs=params['epochs'],
+                batch_size=batch_size,
+                val_split=val_split,
+            )
+            best_epoch = model_checkpoint.best_epoch
+            trainer.model.load_state_dict(torch.load(model_checkpoint.old_files[-1]))
+
+        self.model = trainer
+
+        # TODO: add dynamic epochs selection
+        self.params_trained['epochs'] = params['epochs']
+        self.params_trained['best_epoch'] = best_epoch
+
+    def _predict_proba(self, X, **kwargs):
+        X = self.preprocess(X, **kwargs)
+        X = self.missing_filler.transform(X)
+        X_tab = self._tab_preprocessor.transform(X)
+        if self.problem_type != REGRESSION:
+            preds = self.model.predict_proba(X_tab=X_tab)
+        else:
+            preds = self.model.predict(X_tab=X_tab)
+            preds = self.y_scaler.inverse_transform(preds)
+
+        if self.problem_type == BINARY:
+            return preds[:, 1]
+        else:
+            return preds
+
+    def __get_batch_size(self, params):
+        batch_size = params['bs']
+        # SAINT need larger batches because it is using information between rows
+        if params['type'] == 'SAINT':
+            batch_size *= 2
+        return batch_size
+
+    def __prepare_datasets_before_fit(self, X, y, X_val, y_val):
+        from pytorch_widedeep.preprocessing import TabPreprocessor
+
+        # prepare wide, crossed, embedding and continuous columns
+        cont_cols = self._feature_metadata.get_features(valid_raw_types=[R_INT, R_FLOAT, R_DATETIME])
+
+        cat_cols = self._feature_metadata.get_features(valid_raw_types=[R_OBJECT, R_CATEGORY, R_BOOL])
+        cat_cols = CategoricalFeaturesFilter.filter(X, cat_cols, self.params.get('max_unique_categorical_values', 10000))
+
+        X = self.preprocess(X)
+        if X_val is not None:
+            X_val = self.preprocess(X_val)
+
+        self.missing_filler = MissingFiller(self._feature_metadata)
+        X = self.missing_filler.fit_transform(X)
+        if X_val is not None:
+            X_val = self.missing_filler.transform(X_val)
+
+        self.y_scaler = TargetScaler(self.problem_type, self.params.get('y_scaler', None))
+        y_norm, y_val_norm = self.y_scaler.fit_transform(y, y_val)
+
+        embed_cols, for_transformer = self.__get_embedding_columns_metadata(X, cat_cols)
+        self._tab_preprocessor = TabPreprocessor(
+            embed_cols=embed_cols,
+            continuous_cols=cont_cols,
+            for_transformer=for_transformer
+        )
+
+        X_tab = self._tab_preprocessor.fit_transform(X)
+        X_train = {'X_tab': X_tab, 'target': y_norm.values}
+        if X_val is not None and y_val is not None:
+            X_valid = {'X_tab': self._tab_preprocessor.transform(X_val), 'target': y_val_norm.values}
+            val_split = None
+        else:
+            X_valid = None
+            val_split = 0.1
```

Comment:
So refit_full uses 10% of data for validation? This isn't ideal. We might be better off with

```
    def _more_tags(self):
        return {'can_refit_full': False}
```

which will be either very similar or better in model quality without needing to train again.

Reply:
I set it to `can_refit_full: False` for now. I believe it is possible to implement refit_full without validation set, it requires modifying early stopping callback which is not dependent on the validation metric - added TODO to fix later.